src/nsrt_learning/strips_learning/gen_to_spec_learner.py

```diff
@@ -215,17 +215,35 @@ def _backchain_one_pass(
                 pnad, var_to_obj = self._find_best_matching_pnad_and_sub(
                     segment, objects, pnads_for_option)
                 if pnad is not None:
+                    segs_in_pnad = [
+                        datapoint[0] for datapoint in pnad.datastore
+                    ]
+                    # In this case, we potentially want to recompute_datastores
+                    # to see if this will just resolve the issue.
+                    if segment not in segs_in_pnad:
+                        self._recompute_datastores_from_segments(
+                            pnads_for_option)
+                        # Since we're not changing anything aside from the
+                        # PNADs' datastores, we're guaranteed that the result
+                        # of _find_best_matching_pnad_and_sub will stay the
+                        # same.
+                        segs_in_pnad = [
+                            datapoint[0] for datapoint in pnad.datastore
+                        ]
+                if pnad is not None and segment in segs_in_pnad:
                     assert var_to_obj is not None
                     obj_to_var = {v: k for k, v in var_to_obj.items()}
                     assert len(var_to_obj) == len(obj_to_var)
                     ground_op = pnad.op.ground(
                         tuple(var_to_obj[var] for var in pnad.op.parameters))
                     if len(param_opt_to_nec_pnads[option.parent]) == 0:
                         param_opt_to_nec_pnads[option.parent].append(pnad)
-                # If we weren't able to find a substitution (i.e, the above
-                # _find_best_matching call didn't yield a PNAD), we need to
-                # spawn a new PNAD from the most general PNAD to cover
-                # these necessary add effects.
+                # If we weren't able to find a substitution such that the given
+                # segment is in the particular PNAD's datastore (i.e, the above
+                # _find_best_matching call didn't yield a PNAD such that this
+                # PNAD's datastore contains the segment), we need to spawn a
```

Comment:
wait... if we found a substitution with a PNAD that the segment is NOT in the datastore of, then why should the correct move be to spawn an entirely new PNAD? why wouldn't we just move the segment to the datastore of the PNAD it unified with?